Import statsmodels for linear regression and padas to read the data

In [1]:
import statsmodels.api as sm
import pandas as pd

create dataframe using pandas

In [2]:
df=pd.read_csv('boston.csv')

reading first 5 rows of the data frame

In [4]:
df.head()

CRIM  INDUS    NOX     RM        AGE     DIS  TAX         PT  \
0  0.00632   2.31  0.538  6.575  65.199997  4.0900  296  15.300000   
1  0.02731   7.07  0.469  6.421  78.900002  4.9671  242  17.799999   
2  0.02729   7.07  0.469  7.185  61.099998  4.9671  242  17.799999   
3  0.03237   2.18  0.458  6.998  45.799999  6.0622  222  18.700001   
4  0.06905   2.18  0.458  7.147  54.200001  6.0622  222  18.700001   

            B         MV  
0  396.899994  24.000000  
1  396.899994  21.600000  
2  392.829987  34.700001  
3  394.630005  33.400002  
4  396.899994  36.200001

There are 506 rows and 10 columns in the dataset

In [5]:
df.shape

(506, 10)

Using sklearn to divid data into train and test set

In [6]:
from sklearn.model_selection import train_test_split

In [7]:
X_train,X_test,Y_train,Y_test = train_test_split(df.drop('MV',axis=1),df['MV'],test_size=0.2,random_state=111)

Add constant to the training data (X_train)

In [8]:
X_train=sm.add_constant(X_train)

Fit the OLS model

In [9]:
model=sm.OLS(Y_train,X_train)

In [10]:
result=model.fit()

Print model result/summary

In [11]:
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                     MV   R-squared:                       0.663
Model:                            OLS   Adj. R-squared:                  0.655
Method:                 Least Squares   F-statistic:                     86.16
Date:                Thu, 14 Mar 2019   Prob (F-statistic):           1.91e-87
Time:                        21:41:08   Log-Likelihood:                -1245.4
No. Observations:                 404   AIC:                             2511.
Df Residuals:                     394   BIC:                             2551.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         11.3200      5.980      1.893      0.0

- From the model, the parameter fro CRIM (CRIM per capita crime rate by town), INDUS (INDUS proportion of non-retail business acres per town), and TAX(TAX full-value property-tax rate per $10,000) are not significantly different from zero at 5 percent significance level. 
-Hence we are going to drop predictor with the highest P value, and re-run model.

We going to drop TAX variable to predict MV

In [12]:
X_train=X_train.drop(['TAX'], axis=1)

Fit the updated model

In [13]:
updated_model_result = sm.OLS(Y_train,X_train).fit()

In [14]:
print(updated_model_result.summary())

                            OLS Regression Results                            
Dep. Variable:                     MV   R-squared:                       0.663
Model:                            OLS   Adj. R-squared:                  0.656
Method:                 Least Squares   F-statistic:                     97.17
Date:                Thu, 14 Mar 2019   Prob (F-statistic):           1.89e-88
Time:                        21:43:38   Log-Likelihood:                -1245.5
No. Observations:                 404   AIC:                             2509.
Df Residuals:                     395   BIC:                             2545.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         11.1972      5.825      1.922      0.0

Predictors CRIM and INDUS still have a high P value in the updated model. The parameter fro CRIM and INDUS are not significantly different from zero at 5% significane level. We are going to drop INDUS and re-run the model again since it has the highest P value

In [15]:
X_train=X_train.drop(['INDUS'], axis=1)

Fit the updated model

In [16]:
updated_model_result2 = sm.OLS(Y_train,X_train).fit()

In [17]:
print(updated_model_result2.summary())

                            OLS Regression Results                            
Dep. Variable:                     MV   R-squared:                       0.662
Model:                            OLS   Adj. R-squared:                  0.656
Method:                 Least Squares   F-statistic:                     110.9
Date:                Thu, 14 Mar 2019   Prob (F-statistic):           2.78e-89
Time:                        21:46:13   Log-Likelihood:                -1245.9
No. Observations:                 404   AIC:                             2508.
Df Residuals:                     396   BIC:                             2540.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         11.4237      5.821      1.963      0.0

The parameter fro CRIM is not significantly different from zero at 5% significane level. We are going to drop CRIM and re-run the model.

In [18]:
X_train=X_train.drop(['CRIM'], axis=1)

Fit the updated model

In [19]:
updated_model_result3 = sm.OLS(Y_train,X_train).fit()

In [20]:
print(updated_model_result3.summary())

                            OLS Regression Results                            
Dep. Variable:                     MV   R-squared:                       0.660
Model:                            OLS   Adj. R-squared:                  0.655
Method:                 Least Squares   F-statistic:                     128.3
Date:                Thu, 14 Mar 2019   Prob (F-statistic):           1.04e-89
Time:                        21:48:52   Log-Likelihood:                -1247.5
No. Observations:                 404   AIC:                             2509.
Df Residuals:                     397   BIC:                             2537.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         11.1788      5.833      1.916      0.0

The R-squared value has slighjtly decreased from .663 to .660 and Adjusted R-squared remain the same at 0.655

Import sklearn.metrics to see model's accurancy

In [21]:
from sklearn.metrics import r2_score

Drop the three variables in the X_test as X_train

In [22]:
X_test=X_test.drop(['CRIM','INDUS','TAX'], axis=1)

Make preduction on the test set

In [24]:
predictions = updated_model_result3.predict(sm.add_constant(X_test))

Check R-squared value

In [25]:
r2_score(Y_test, predictions)

0.621648074202985

Checking sample values from the actual and predictions

In [28]:
pd.DataFrame({'actual value': Y_test, 'predictions':predictions}).sample(10)

actual value  predictions
497     18.299999    19.291337
250     24.400000    25.243696
385      7.200000    12.557819
405      5.000000    15.197410
107     20.400000    20.762280
146     15.600000    13.799426
149     15.400000    17.096466
409     27.500000    21.412902
286     20.100000    21.313205
39      30.799999    28.688188